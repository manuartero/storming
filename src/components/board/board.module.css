.board {
  --tile-size: 12vh;
  --collapse-rows-y: calc(var(--tile-size) / -5);
  --row-left-space: calc(var(--tile-size) / 2);

  /** main grid layout */
  grid-area: board;

  position: relative;
  width: -moz-fit-content;
  width: fit-content;
  height: -moz-fit-content;
  height: fit-content;
}

.row {
  display: flex;
  position: relative;

  width: -moz-fit-content;
  width: fit-content;
}

.row:nth-child(2),
.row:nth-child(3),
.row:nth-child(4),
.row:nth-child(5),
.row:nth-child(6),
.row:nth-child(7) {
  margin-top: var(--collapse-rows-y);
}

.row.row1ToEquator {
  left: var(--row-left-space);
}

.row.row2ToEquator {
  left: calc(var(--row-left-space) * 2);
}

.row.row3ToEquator {
  left: calc(var(--row-left-space) * 3);
}
