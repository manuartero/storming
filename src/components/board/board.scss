$TILE_SIZE: 12vh;
$collapse-rows-y: calc($TILE_SIZE / 5); // 22px aprox
$row-left-space: calc($TILE_SIZE / 2);

.board {
  position: relative;

  width: -moz-fit-content;
  width: fit-content;
  height: -moz-fit-content;
  height: fit-content;

  &__row {
    display: flex;
    position: relative;

    width: -moz-fit-content;
    width: fit-content;

    // Loop for collapsing rows
    @for $i from 2 through 7 {
      &:nth-child(#{$i}) {
        margin-top: $collapse-rows-y * (-1);
      }
    }

    // Loop for moving rows to the right
    @for $i from 1 through 3 {
      &--#{$i}-to-equator {
        left: $row-left-space * $i;
      }
    }
  }

  // debug
  // border: 1px solid red;
}
