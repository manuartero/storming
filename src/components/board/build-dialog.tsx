import { CardSilhouette } from "components/cards";
import { Dialog } from "elements";

type Props = {
  onWallOption: () => void;
  onUpgradeOption: () => void;
  onClose: () => void;
};

export function BuildDialog({ onWallOption, onUpgradeOption, onClose }: Props) {
  return (
    <Dialog
      size="small"
      title="Choose How to Improve your Building"
      onClose={onClose}
    >
      <>
        <CardSilhouette card="build-walls" onClick={onWallOption} />
        <CardSilhouette card="upgrade-settlement" onClick={onUpgradeOption} />
      </>
    </Dialog>
  );
}
