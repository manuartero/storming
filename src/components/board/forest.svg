<svg version="1.0" xmlns="http://www.w3.org/2000/svg"
 viewBox="0 0 512.000000 512.000000"
 preserveAspectRatio="xMidYMid meet">

<g transform="translate(0.000000,512.000000) scale(0.100000,-0.100000)">
<path d="M1002 4518 c-78 -39 -184 -135 -234 -212 -21 -33 -50 -96 -65 -140
l-27 -81 -4 -406 -4 -406 -63 -42 c-204 -133 -417 -521 -525 -951 -57 -231
-65 -295 -65 -565 0 -226 2 -258 23 -339 33 -129 75 -210 146 -282 100 -99
207 -137 426 -150 l110 -7 0 -108 0 -108 -341 -3 -341 -3 -19 -24 c-25 -30
-24 -62 2 -95 l20 -26 2515 0 c2501 0 2514 0 2534 20 30 30 27 83 -6 109 -25
19 -40 21 -206 21 l-179 0 3 133 3 132 71 6 c166 15 261 104 294 276 18 91 8
324 -18 439 -26 111 -63 216 -108 307 -19 38 -31 74 -27 80 4 7 26 36 50 64
50 62 105 173 129 263 28 101 26 292 -5 390 -59 192 -177 343 -342 441 -66 40
-199 89 -239 89 -13 0 -32 20 -55 55 -48 75 -105 136 -170 181 -38 25 -55 44
-55 59 0 39 -31 140 -64 210 -159 334 -554 504 -901 387 l-53 -18 -52 73
c-147 213 -317 171 -457 -112 -81 -162 -167 -435 -224 -706 l-32 -156 -74 34
c-126 58 -210 76 -365 77 l-138 1 0 316 c0 283 -2 324 -20 394 -27 104 -79
195 -159 275 -72 71 -190 140 -242 140 -39 0 -69 -32 -69 -73 0 -40 13 -53 95
-93 107 -53 199 -166 231 -288 10 -38 14 -134 14 -382 l0 -332 -75 -36 c-99
-46 -167 -92 -237 -160 l-58 -55 0 62 0 62 80 80 c89 89 101 122 57 163 -31
29 -69 28 -106 -4 l-31 -26 0 237 0 236 -25 24 c-31 32 -74 33 -103 2 -18 -19
-22 -35 -22 -92 l0 -68 -50 47 c-35 33 -57 46 -78 46 -43 0 -75 -30 -75 -69 0
-28 15 -47 101 -134 l102 -102 0 -290 -1 -290 -31 40 c-98 125 -223 220 -315
240 l-33 7 0 346 c0 191 4 373 10 406 23 140 116 266 246 333 81 42 102 70 85
116 -21 55 -69 60 -159 15z m2004 -275 c40 -43 98 -148 139 -253 43 -109 62
-133 104 -133 69 0 88 59 51 153 -11 28 -20 55 -20 60 0 15 99 40 180 46 224
17 454 -112 557 -312 22 -44 43 -90 46 -104 l6 -25 -110 3 c-97 3 -121 1 -192
-22 -100 -31 -163 -69 -243 -148 -34 -33 -64 -56 -66 -52 -3 5 -13 44 -22 86
-13 59 -23 82 -43 98 -30 23 -42 24 -77 9 -44 -20 -45 -50 -10 -220 76 -359
108 -675 108 -1049 -1 -610 -84 -820 -326 -820 l-48 0 0 192 0 193 129 130
c71 71 130 137 131 145 7 53 -23 90 -71 90 -25 0 -46 -14 -110 -77 l-79 -76 0
320 c0 309 -1 321 -21 347 -15 19 -30 26 -54 26 -56 0 -75 -35 -75 -137 l0
-85 -65 61 c-72 67 -96 74 -139 40 -19 -15 -26 -29 -26 -56 0 -33 11 -48 115
-153 l115 -116 0 -423 0 -424 -67 5 c-73 5 -133 30 -170 70 -196 210 -185
1217 22 2013 86 331 223 625 290 625 9 0 27 -12 41 -27z m1087 -736 c121 -44
200 -119 260 -244 l32 -68 3 -522 3 -522 -24 -38 c-74 -117 -137 -284 -168
-446 -11 -56 -25 -107 -32 -114 -14 -14 -112 -53 -132 -53 -13 0 -15 33 -15
217 l0 217 80 82 c70 71 80 86 80 117 0 26 -7 41 -26 56 -34 27 -53 26 -97 -3
l-36 -25 -3 241 c-3 238 -3 240 -27 259 -31 25 -69 24 -98 -3 -20 -18 -23 -31
-23 -91 l0 -69 -51 46 c-55 49 -88 58 -126 31 -28 -20 -37 -59 -19 -93 8 -15
55 -67 105 -117 l91 -90 0 -389 0 -388 -32 6 c-110 22 -235 118 -292 225 l-29
54 17 108 c35 218 44 494 26 774 -10 156 -34 384 -50 476 -8 50 -7 59 26 126
56 117 157 207 274 244 82 27 203 25 283 -4z m-1908 -252 c33 -8 98 -34 145
-58 l85 -42 -2 -40 c-2 -22 -10 -98 -18 -170 -29 -252 -40 -547 -27 -769 6
-112 14 -210 17 -217 4 -10 -15 -25 -60 -47 -46 -22 -67 -28 -69 -20 -2 7 -11
55 -20 105 -22 123 -58 244 -112 380 -40 100 -44 119 -44 194 l0 83 90 92 c75
76 90 97 90 123 0 41 -31 71 -72 71 -24 0 -42 -10 -70 -37 l-38 -37 0 59 c0
47 -5 64 -22 82 -26 28 -67 30 -99 4 -23 -18 -24 -27 -27 -150 l-4 -131 -56
50 c-81 72 -133 95 -212 95 -84 -1 -134 -27 -221 -114 l-67 -67 -21 43 c-12
24 -21 51 -21 59 0 9 19 53 42 97 49 96 177 231 270 285 65 39 172 78 249 91
70 13 218 6 294 -14z m2451 -115 c98 -49 168 -109 228 -197 143 -210 138 -486
-14 -692 l-32 -43 -31 30 c-66 62 -155 87 -221 62 l-26 -9 -2 444 c-2 327 0
445 8 445 7 0 47 -18 90 -40z m-3767 -12 c101 -49 228 -211 332 -425 75 -155
78 -167 45 -231 -45 -89 -86 -201 -86 -237 0 -24 7 -39 26 -54 60 -47 98 -14
154 131 68 177 173 335 259 390 57 36 77 35 133 -2 128 -85 278 -374 341 -660
62 -276 58 -528 -8 -660 -45 -88 -101 -118 -236 -127 l-89 -6 0 114 0 113 130
131 c118 120 130 135 130 168 0 39 -36 77 -73 77 -11 0 -55 -35 -103 -82 l-84
-82 0 198 c0 196 0 197 -25 221 -31 32 -69 32 -100 0 l-25 -24 0 -417 0 -417
-87 5 c-49 3 -107 13 -131 23 -121 49 -171 205 -157 494 6 139 6 144 -16 167
-36 39 -90 31 -114 -17 -22 -41 -31 -271 -16 -384 24 -177 67 -271 159 -347
l33 -27 -30 -16 c-52 -26 -137 -44 -251 -51 l-110 -7 0 393 0 393 70 73 c39
39 70 79 70 87 0 32 -23 66 -53 78 -25 11 -34 10 -58 -4 l-29 -16 0 115 c0
110 -1 116 -26 141 -20 20 -31 24 -57 19 -67 -13 -67 -11 -67 -379 l0 -331
-47 46 c-78 75 -108 84 -148 43 -16 -15 -25 -35 -25 -56 0 -27 17 -48 110
-143 l110 -111 0 -172 0 -173 -47 0 c-87 0 -222 20 -276 40 -106 40 -165 115
-209 265 -21 74 -23 99 -23 320 1 268 15 378 81 608 110 384 310 724 474 805
57 28 92 28 149 0z m3836 -1020 c89 -93 174 -288 210 -482 24 -129 18 -333
-12 -399 -26 -57 -80 -86 -155 -87 l-48 0 0 62 0 62 75 77 c82 84 90 107 54
153 -23 30 -78 36 -99 11 -7 -8 -16 -15 -21 -15 -5 0 -9 54 -9 120 0 107 -2
122 -20 140 -25 25 -71 26 -101 1 l-24 -19 -3 -297 -3 -298 -61 6 c-102 9
-139 54 -159 192 -26 185 37 473 145 665 41 74 125 160 156 160 16 0 41 -18
75 -52z m-2226 -486 c49 -96 139 -169 241 -196 l55 -14 -66 -31 c-92 -44 -196
-64 -287 -57 -75 6 -178 31 -187 45 -2 5 2 39 10 77 8 38 17 117 21 176 l7
107 46 16 c25 9 57 22 71 30 25 13 25 13 46 -48 12 -34 31 -82 43 -105z m1015
-80 l36 -44 0 -389 0 -389 -85 0 -85 0 0 391 0 390 42 49 c24 27 46 46 49 42
4 -4 23 -27 43 -50z m-284 -472 l0 -350 -85 0 -85 0 0 345 0 344 63 4 c34 2
72 4 85 5 l22 2 0 -350z m600 286 l55 -13 3 -311 2 -312 -95 0 -95 0 0 340 0
340 38 -16 c20 -8 62 -21 92 -28z m372 -48 c22 -194 126 -303 301 -316 l67 -5
0 -133 0 -134 -265 0 -265 0 0 313 0 312 63 16 c91 24 91 24 99 -53z m-1292
-293 l0 -295 -150 0 -150 0 2 231 3 232 60 16 c64 18 177 69 210 95 11 8 21
16 23 16 1 0 2 -133 2 -295z m-532 165 l82 -12 0 -224 0 -224 -181 0 -182 0 5
225 4 225 33 34 32 33 63 -23 c34 -12 99 -28 144 -34z m-837 -80 l69 0 0 -190
0 -190 -120 0 -120 0 0 157 0 157 37 38 c21 21 44 36 51 33 8 -3 45 -5 83 -5z
m409 -185 l0 -195 -95 0 -95 0 0 190 0 189 78 4 c42 1 85 4 95 5 16 2 17 -13
17 -193z m-730 -70 l0 -125 -165 0 -165 0 0 109 0 108 113 8 c61 3 130 10 152
15 22 4 46 8 53 9 9 1 12 -30 12 -124z"/>
</g>
</svg>
