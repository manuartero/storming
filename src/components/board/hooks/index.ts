import { useReducer } from "react";
import boardReducer from "./board-reducer";
import initialBoard from "./initial-board";

export function useBoard() {
  const [boardState, dispatchBoardAction] = useReducer(
    boardReducer,
    initialBoard
  );
  return { boardState, dispatchBoardAction };
}
