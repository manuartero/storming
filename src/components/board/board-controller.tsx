import { useGameContext } from "game-context";
import { NewBuilding, upgradeBuilding } from "models/new-building";
import { useState } from "react";
import { warnInconsistentState } from "utils/console";
import { Board } from "./board";
import { BuildDialog } from "./build-dialog";
import { inferVisualBoardFromGameContext } from "./infer-visual-board";
import { RecruitDialog } from "./recruit-dialog";

/**
 * Defines visual board from GameContext:
 *  - selected, available and forbidden tiles
 *  - `onTileClick()`
 *
 *  => state validation before changing GameContext <=
 */
export function BoardController() {
  const gameContext = useGameContext();
  const [selectedTile, setSelectedTile] = useState<TileID>();
  const [buildingTile, setBuildingTile] = useState<TileID>();
  const [recruitingTile, setRecruitingTile] = useState<TileID>();

  /* derived state */
  const board = inferVisualBoardFromGameContext(gameContext, selectedTile);

  const settleOnTile = (tile: TileID) => {
    const piece = board[tile].piece;
    if (!piece) {
      return warnInconsistentState(
        `trying to settle on ${tile} but no soldier found`,
        { tile: board[tile] }
      );
    }
    setSelectedTile(undefined);
    setBuildingTile(undefined);
    setRecruitingTile(undefined);
    gameContext.build({
      tile,
      building: NewBuilding({ owner: piece.owner }),
    });
  };

  const upgradeBuildingOnTile = (tile: TileID) => {
    const building = board[tile].building;
    if (!building) {
      return warnInconsistentState(
        `trying to upgrade building on ${tile} but no building found`,
        { tile: board[tile] }
      );
    }
    setSelectedTile(undefined);
    setBuildingTile(undefined);
    setRecruitingTile(undefined);

    gameContext.build({
      tile,
      building: upgradeBuilding(building),
    });
  };

  const buildWallsOnTile = (tile: TileID) => {
    const building = board[tile].building;
    if (!building) {
      return warnInconsistentState(
        `trying to build walls on ${tile} but no building found`,
        { tile: board[tile] }
      );
    }
    setSelectedTile(undefined);
    setBuildingTile(undefined);
    setRecruitingTile(undefined);
    gameContext.build({
      tile,
      building: {
        ...building,
        hasWalls: true,
      },
    });
  };

  const discardOptionDialog = () => {
    setSelectedTile(undefined);
    setBuildingTile(undefined);
    setRecruitingTile(undefined);
  };

  const moveFromTile = (tile: TileID) => {
    if (selectedTile) {
      const piece = board[selectedTile].piece;
      if (!piece) {
        return setSelectedTile(tile);
      }
      if (piece.owner === gameContext.activePlayer) {
        setSelectedTile(undefined);
        setBuildingTile(undefined);
        setRecruitingTile(undefined);
        return gameContext.move({
          piece,
          from: selectedTile,
          to: tile,
        });
      }
    }
    setSelectedTile(tile);
  };

  const recruitOnTile = (tile: TileID, piece: PieceType) => {
    const building = board[tile].building;
    if (!building) {
      return warnInconsistentState(
        `trying to recruit ${piece} on ${tile} but no building found`,
        { tile: board[tile] }
      );
    }
    setSelectedTile(undefined);
    setBuildingTile(undefined);
    setRecruitingTile(undefined);
    return gameContext.recruit({
      tile,
      piece: {
        type: piece,
        owner: building.owner,
      },
    });
  };

  const resolveRecruitOnTile = (tile: TileID) => {
    const building = board[tile].building;
    if (!building) {
      return warnInconsistentState(
        `trying to recruit on ${tile} but no building found`,
        { tile: board[tile] }
      );
    }
    if (board[tile].piece) {
      return warnInconsistentState(
        `trying to recruit on ${tile} but piece already found`,
        { tile: board[tile] }
      );
    }
    setRecruitingTile(tile);
  };

  const resolveBuildOnTile = (tile: TileID) => {
    if (!board[tile].building && board[tile].piece?.type === "soldier") {
      return settleOnTile(tile);
    }
    if (board[tile].building?.hasWalls) {
      return upgradeBuildingOnTile(tile);
    }
    setBuildingTile(tile);
  };

  const resolveActionOnTile = (tile: TileID) => {
    if (gameContext.activeCard?.cardType !== "actionCard") {
      return warnInconsistentState(
        `trying to resolve an action on ${tile} while no action card`,
        { tile: board[tile] }
      );
    }
    switch (gameContext.activeCard.action) {
      case "build":
        return resolveBuildOnTile(tile);
      case "move":
        return moveFromTile(tile);
      case "recruit":
        return resolveRecruitOnTile(tile);
    }
  };

  const onTileClick = ({ str: tile }: Coordinates) => {
    if (board[tile].status === "available") {
      return resolveActionOnTile(tile);
    }
    setSelectedTile(tile);
  };

  return (
    <div className="board-container">
      <Board
        state={board}
        activePlayer={gameContext.activePlayer}
        onTileClick={onTileClick}
      />
      {buildingTile && (
        <BuildDialog
          onWallOption={() => {
            buildWallsOnTile(buildingTile);
          }}
          onUpgradeOption={() => {
            upgradeBuildingOnTile(buildingTile);
          }}
          onClose={() => {
            discardOptionDialog();
          }}
        />
      )}
      {recruitingTile && (
        <RecruitDialog
          onSoldierOption={() => {
            recruitOnTile(recruitingTile, "soldier");
          }}
          onKnightOption={() => {
            recruitOnTile(recruitingTile, "knight");
          }}
          onClose={() => {
            discardOptionDialog();
          }}
        />
      )}
    </div>
  );
}
