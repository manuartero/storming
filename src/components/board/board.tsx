import { row } from "models/tiles";
import Piece from "../pieces/piece";
import Tile from "./tile";

import "./board.scss";

interface Props {
  state: VisualBoard;
  onTileClick: (titleID: Coordinates) => void;
}

function Board({ state, onTileClick }: Props): JSX.Element {

  const renderRow = (n: -3 | -2 | -1 | 0 | 1 | 2 | 3) =>
    row(n).map((tileId) => {
      const s = state[tileId];
      return (
        <Tile
          key={tileId}
          id={tileId}
          terrain={s.terrain}
          building={s.building?.type}
          owner={s.building?.owner}
          status={s.status}
          onClick={onTileClick}
        >
          {s.piece && <Piece type={s.piece.type} owner={s.piece.owner} />}
        </Tile>
      );
    });

  return (
    <div className="board">
      <div className="board__row board__row--3-to-equator">{renderRow(-3)}</div>
      <div className="board__row board__row--2-to-equator">{renderRow(-2)}</div>
      <div className="board__row board__row--1-to-equator">{renderRow(-1)}</div>
      <div className="board__row">{renderRow(0)}</div>
      <div className="board__row board__row--1-to-equator">{renderRow(1)}</div>
      <div className="board__row board__row--2-to-equator">{renderRow(2)}</div>
      <div className="board__row board__row--3-to-equator">{renderRow(3)}</div>
    </div>
  );
}

export default Board;
