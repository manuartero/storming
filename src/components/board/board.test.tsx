import { render } from "@testing-library/react";
import { Board } from "./board";

describe("<Board />", () => {
  test("render: match snapshot", () => {
    const { asFragment } = render(
      <Board state={visualBoard} onTileClick={jest.fn()} />
    );
    expect(asFragment()).toMatchSnapshot();
  });
});

const visualBoard: VisualBoard = {
  "-2,-3": {
    status: "available",
  },
  "-1,-3": {
    status: "available",
  },
  "0,-3": {
    building: { owner: "enemy1", type: "castle" },
    status: "available",
  },
  "1,-3": {
    status: "available",
  },
  "2,-3": {
    status: "available",
  },
  "-3,-2": {
    status: "available",
  },
  "-2,-2": {
    status: "available",
  },
  "-1,-2": {
    piece: { owner: "enemy1", type: "soldier" },
    status: "available",
  },
  "0,-2": {
    status: "available",
  },
  "1,-2": {
    status: "available",
  },
  "2,-2": {
    status: "available",
  },
  "-3,-1": {
    status: "available",
  },
  "-2,-1": {
    status: "available",
  },
  "-1,-1": {
    terrain: "lake",
    status: "available",
  },
  "0,-1": {
    status: "available",
  },
  "1,-1": {
    status: "available",
  },
  "2,-1": {
    status: "available",
  },
  "3,-1": {
    status: "available",
  },
  "-4,0": {
    building: { owner: "player", type: "castle" },
    status: "available",
  },
  "-3,0": {
    piece: { owner: "player", type: "soldier" },
    status: "available",
  },
  "-2,0": {
    status: "available",
  },
  "-1,0": {
    status: "available",
  },
  "0,0": {
    status: "available",
  },
  "1,0": {
    status: "available",
  },
  "2,0": {
    piece: { owner: "enemy2", type: "soldier" },
    status: "available",
  },
  "3,0": {
    building: { owner: "enemy2", type: "castle" },
    status: "available",
  },
  "-3,1": {
    status: "available",
  },
  "-2,1": {
    status: "available",
  },
  "-1,1": {
    status: "available",
  },
  "0,1": {
    status: "available",
  },
  "1,1": {
    terrain: "lake",
    status: "available",
  },
  "2,1": {
    status: "available",
  },
  "3,1": {
    status: "available",
  },
  "-3,2": {
    status: "available",
  },
  "-2,2": {
    status: "available",
  },
  "-1,2": {
    status: "available",
  },
  "0,2": {
    piece: { owner: "enemy3", type: "soldier" },
    status: "available",
  },
  "1,2": {
    status: "available",
  },
  "2,2": {
    status: "available",
  },
  "-2,3": {
    status: "available",
  },
  "-1,3": {
    status: "available",
  },
  "0,3": {
    building: { owner: "enemy3", type: "castle" },
    status: "available",
  },
  "1,3": {
    status: "available",
  },
  "2,3": {
    status: "available",
  },
};
