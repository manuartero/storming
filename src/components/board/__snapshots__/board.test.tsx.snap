// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`<Board /> render: match snapshot 1`] = `
<DocumentFragment>
  <div
    class="board"
  >
    <div
      class="board__row board__row--3-to-equator"
    >
      <button
        aria-disabled="false"
        aria-label="tile -2,-3"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -1,-3"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 0,-3"
        class="tile available enemy1"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
        <div
          aria-label="enemy1 castle"
          aria-roledescription="game building"
          class="building"
          role="img"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 1,-3"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 2,-3"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
    </div>
    <div
      class="board__row board__row--2-to-equator"
    >
      <button
        aria-disabled="false"
        aria-label="tile -3,-2"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -2,-2"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -1,-2"
        class="tile available enemy1"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        >
          <div
            class="content"
          >
            <div
              aria-label="enemy1 soldier"
              aria-roledescription="game piece"
              class="piece ignore-clicks"
              role="img"
            />
          </div>
        </div>
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 0,-2"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 1,-2"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 2,-2"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
    </div>
    <div
      class="board__row board__row--1-to-equator"
    >
      <button
        aria-disabled="false"
        aria-label="tile -3,-1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -2,-1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -1,-1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        >
          <div
            aria-label="terrain lake"
            aria-roledescription="game terrain"
            class="terrain"
            role="img"
          />
        </div>
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 0,-1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 1,-1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 2,-1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 3,-1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
    </div>
    <div
      class="board__row"
    >
      <button
        aria-disabled="false"
        aria-label="tile -4,0"
        class="tile available player"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
        <div
          aria-label="player castle"
          aria-roledescription="game building"
          class="building"
          role="img"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -3,0"
        class="tile available player"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        >
          <div
            class="content"
          >
            <div
              aria-label="player soldier"
              aria-roledescription="game piece"
              class="piece ignore-clicks"
              role="img"
            />
          </div>
        </div>
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -2,0"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -1,0"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 0,0"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 1,0"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 2,0"
        class="tile available enemy2"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        >
          <div
            class="content"
          >
            <div
              aria-label="enemy2 soldier"
              aria-roledescription="game piece"
              class="piece ignore-clicks"
              role="img"
            />
          </div>
        </div>
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 3,0"
        class="tile available enemy2"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
        <div
          aria-label="enemy2 castle"
          aria-roledescription="game building"
          class="building"
          role="img"
        />
      </button>
    </div>
    <div
      class="board__row board__row--1-to-equator"
    >
      <button
        aria-disabled="false"
        aria-label="tile -3,1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -2,1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -1,1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 0,1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 1,1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        >
          <div
            aria-label="terrain lake"
            aria-roledescription="game terrain"
            class="terrain"
            role="img"
          />
        </div>
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 2,1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 3,1"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
    </div>
    <div
      class="board__row board__row--2-to-equator"
    >
      <button
        aria-disabled="false"
        aria-label="tile -3,2"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -2,2"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -1,2"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 0,2"
        class="tile available enemy3"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        >
          <div
            class="content"
          >
            <div
              aria-label="enemy3 soldier"
              aria-roledescription="game piece"
              class="piece ignore-clicks"
              role="img"
            />
          </div>
        </div>
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 1,2"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 2,2"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
    </div>
    <div
      class="board__row board__row--3-to-equator"
    >
      <button
        aria-disabled="false"
        aria-label="tile -2,3"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile -1,3"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 0,3"
        class="tile available enemy3"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
        <div
          aria-label="enemy3 castle"
          aria-roledescription="game building"
          class="building"
          role="img"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 1,3"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
      <button
        aria-disabled="false"
        aria-label="tile 2,3"
        class="tile available"
        type="button"
      >
        <div
          class="strokeLayer"
        />
        <div
          class="innerLayer"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
